### PR TITLE
[FIX] sale,website_sale: fix currency conversion of product

### LINKED ISSUE #29905: Multi-website, Multi-currency, Multi-company and multi-pricelist on website are throwing an error 500
This line https://github.com/odoo/odoo/commit/f8dad1bb4b712d8f34035420ef1f6e484c638f5b#diff-349b8776d2f59616d627bfa5dc90cf49R261

Causes this error:

What are the steps to reproduce your issue?
- Create a website "website_2"
- Create a company "company_2"
- Select any published product and:
 a) set the website to  "website_2"
 b) Leave the company field empty.
- Select "company_2" and:
a) set website to "website_2"
b) set an URL to that company-website combo
- Select two pricelists and:
a) one set the currency in USD and website to "website_2"
b) other set the currency in EUR and website to "website_2"
- go to the URL set in incognito mode and verify you are in website_2 and company_2
- Go to /shop and see the error 500
What is the current behavior that you observe?
The website is trying to get the company from the product but the product has no company.

What would be your expected behavior in this case?
If  I need to manage a product across several companies I need to be able to leave the company field empty in any product.
                                                    
Video: https://www.youtube.com/watch?v=IKURqK7SAI8

### PR DESCRIPTION
**Pull request by seb-odoo**
In commit f8dad1b, we tried to fix the currency
conversion of products by always using the currency of the product as the base.
This is still correct.

However, the conversion method also needs to know the company (to find the
correct currency rate), and it was assumed the company was also always set on
the product. This is of course not correct when using multi-company.

In this commit we fix that issue by falling back if no company is set on the
product. The order of the fallback is: product website company, given website
company, given pricelist company, current user company.

closes #29905
closes #29946
closes #30046

Co-authored-by: Sébastien Theys <seb@odoo.com>
Co-authored-by: Wolfgang Taferner <wolfgang.taferner@wt-io-it.at>

### CODE REVIEW

**Comment by seb-odoo** [comment]
@JKE-be priority task, this fixes a traceback in stable. runbot red but it's false positive so we can review

**Comment by seb-odoo** [comment]
Rebased and runbot issue has been fixed, this is green now, we can review/merge.

**Comment by seb-odoo** [comment]
@JKE-be reworked PR to add all possible fallbacks. can be reviewed again

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -150,9 +150,10 @@ def _get_pricelist_context(self):
         return pricelist_context, pricelist
 
     def _get_compute_currency(self, pricelist, product=None):
-        from_currency = (product or request.env.user.company_id).currency_id
+        company = product and product._get_current_company(pricelist=pricelist, website=request.website) or pricelist.company_id or request.website.company_id
+        from_currency = (product or company).currency_id
```
Comment (seb-odoo) [inline addons/website_sale/controllers/main.py]: this line is not correct, from currency should not be contextual

**Comment by seb-odoo** [comment]
to rebase with ff

File: addons/website_sale/controllers/main.py [hunk 2]
```diff
@@ -150,9 +150,10 @@ def _get_pricelist_context(self):
         return pricelist_context, pricelist
 
     def _get_compute_currency(self, pricelist, product=None):
-        from_currency = (product or request.env.user.company_id).currency_id
+        company = product and product._get_current_company(pricelist=pricelist, website=request.website) or pricelist.company_id or request.website.company_id
+        from_currency = (product or request.env['res.company']._get_main_company()).currency_id
```
Comment (pedrobaeza) [inline addons/website_sale/controllers/main.py]: Shouldn't this be website company?

Comment (seb-odoo) [reply addons/website_sale/controllers/main.py]: As explained on the 2nd commit, when using multi-company, the base currency with which all prices are encoded is the one from the main company, and not contextual.

What is contextual is the currency rate.

Comment (pedrobaeza) [reply addons/website_sale/controllers/main.py]: OK, thanks for explaining.

Comment (seb-odoo) [reply addons/website_sale/controllers/main.py]: You're welcome. If you want the full details, the only place where this method is used is list view of variants to convert the `lst_price` of the variant (encoded in base currency) to the website currency.

Eventually I want to get rid of that method `_get_compute_currency` all together but this is for master, in all the other places in 12.0 except that one it has already been replaced by `get_combination_info`.

Comment (pedrobaeza) [reply addons/website_sale/controllers/main.py]: Thanks also for the expanded details!! All the price computation through pricelist requires a bit refactoring IMO. It's hardly extendable (look for example the ugly hacks for getting some things in https://github.com/OCA/product-attribute/tree/11.0/product_pricelist_supplierinfo).

Ping me if you want me to review anything for master regarding this.

**Comment by JKE-be** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**